op/aaudio.c

```diff
@@ -0,0 +1,891 @@
+// number of channels)
+static aaudio_channel_mask_t cmus_channel_map_to_aaudio_mask(int channels, const channel_position_t *channel_map)
+{
+	aaudio_channel_mask_t mask = 0;
+
+	// we can only convert a valid channel map
+	if (channels >= CHANNELS_MAX || !channel_map || !channel_map_valid(channel_map)) {
+		return AAUDIO_CHANNEL_INVALID;
+	}
+
+	// special case for mono since cmus defines a separate channel position
+	// for it
+	if (channels == 1 && channel_map[0] == CHANNEL_POSITION_MONO) {
+		return AAUDIO_CHANNEL_FRONT_LEFT;
+	}
+
+	// fill the mask, returning invalid if it has duplicates or no mapping
+	for (int i = 0; i < channels; i++) {
+		#define X(aaudio) \
+		if (A2C__##aaudio != CHANNEL_POSITION_INVALID && channel_map[i] == A2C__##aaudio) { \
+			if (mask & AAUDIO_CHANNEL_##aaudio) \
+				return AAUDIO_CHANNEL_INVALID; \
+			mask |= AAUDIO_CHANNEL_##aaudio; \
+		}
+		A2C_CHANNELS
+		#undef X
+	}
+
+	return mask;
+}
+
+// get the expected cmus channel order for the specified aaudio channel mask
+static bool channel_map_init_aaudio(aaudio_channel_mask_t mask, channel_position_t *map)
+{
+	switch (mask) {
+	#define X(aaudio) \
+	case AAUDIO_CHANNEL_##aaudio: channel_map_copy(map, (channel_position_t[CHANNELS_MAX]){ A2C__##aaudio }); return true;
+	A2C_LAYOUTS
+	#undef X
+	}
+	return false;
+}
+
+// get the name of a known aaudio channel mask
+static const char *aaudio_channel_to_string(aaudio_channel_mask_t mask)
+{
+	switch (mask) {
+		#define X(aaudio) \
+		case AAUDIO_CHANNEL_##aaudio: return #aaudio;
+		A2C_CHANNELS
+		#undef X
+	}
+	switch (mask) {
+		#define X(aaudio) \
+		case AAUDIO_CHANNEL_##aaudio: return #aaudio;
+		A2C_LAYOUTS
+		#undef X
+	}
+	return NULL;
+}
+
+// allocate a map of output frame byte indexes to input frame byte indexes (or
+// -1 to zero) to remap channels (map must be sf_get_frame_size elements)
+static ssize_t *make_channel_remap(const channel_position_t *channel_map_out, const channel_position_t *channel_map_in, sample_format_t sf)
+{
+	int byte, channel_out, channel_in;
+	ssize_t *map;
+
+	map = malloc(sizeof(ssize_t) * (size_t) sf_get_frame_size(sf));
+	if (!map) {
+		return NULL;
+	}
+
+	if (!channel_map_out || !channel_map_valid(channel_map_out) || !channel_map_in || !channel_map_valid(channel_map_in)) {
+		for (byte = 0; byte < sf_get_frame_size(sf); byte++) {
+			map[byte] = byte;
+		}
+	} else {
+		for (byte = 0; byte < sf_get_frame_size(sf); byte++) {
+			map[byte] = -1;
+		}
+		for (channel_out = 0; channel_out < sf_get_channels(sf); channel_out++) {
+			if (channel_map_out[channel_out] != CHANNEL_POSITION_INVALID) {
+				for (channel_in = 0; channel_in < sf_get_channels(sf); channel_in++) {
+					if (channel_map_in[channel_in] == channel_map_out[channel_out]) {
+						for (byte = 0; byte < sf_get_sample_size(sf); byte++) {
+							map[sf_get_sample_size(sf) * channel_out + byte] = (ssize_t) sf_get_sample_size(sf) * channel_in + byte;
```

Comment:
My goal with this was to keep the sample copy logic simple since it gets a bit fiddly to write, and there isn't current any helper to do this. I'll probably keep it this way for now, but if I end up needing to remap samples again, I might import miniaudio's code for doing it and refactor cmus to use it.